ambry-router/src/main/java/com.github.ambry.router/PutOperation.java

```diff
@@ -209,25 +286,108 @@ boolean isEncryptionEnabled() {
     return isEncryptionEnabled;
   }
 
+  /**
+   * Either start reading from the channel containing the blob data for normal put requests or assemble the metadata
+   * chunk for stitch requests.
+   */
+  void startOperation() {
+    Exception exception = null;
+    try {
+      if (options.isChunkUpload() && options.getMaxUploadSize() > routerConfig.routerMaxPutChunkSizeBytes) {
+        exception = new RouterException("Invalid max upload size for chunk upload: " + options.getMaxUploadSize(),
+            RouterErrorCode.InvalidPutArgument);
+      } else if (isStitchOperation()) {
+        processChunksToStitch();
+      } else {
+        startReadingFromChannel();
+      }
+    } catch (Exception e) {
+      exception = e;
+    } finally {
+      if (exception != null) {
+        setOperationExceptionAndComplete(exception);
+        routerCallback.onPollReady();
+      }
+    }
+  }
+
   /**
    * Start reading from the channel containing the data for this operation.
    */
-  void startReadingFromChannel() {
-    channel.readInto(chunkFillerChannel, new Callback<Long>() {
-      @Override
-      public void onCompletion(Long result, Exception exception) {
-        if (exception != null) {
-          setOperationExceptionAndComplete(exception);
-          routerCallback.onPollReady();
-        } else {
-          blobSize = result;
-          chunkFillingCompletedSuccessfully = true;
-        }
-        chunkFillerChannel.close();
+  private void startReadingFromChannel() {
+    channel.readInto(chunkFillerChannel, (result, exception) -> {
+      if (exception != null) {
+        setOperationExceptionAndComplete(exception);
+        routerCallback.onPollReady();
+      } else {
+        blobSize = result;
+        chunkFillingCompletedSuccessfully = true;
       }
+      chunkFillerChannel.close();
     });
   }
 
+  /**
+   * Process the list of data chunks provided and prepare a metadata chunk with the provided chunk IDs.
+   * @throws RouterException if validation of the chunks to stitch failed.
+   */
+  private void processChunksToStitch() throws RouterException {
+    if (chunksToStitch.isEmpty()) {
+      // The current metadata format does not support empty chunk lists
+      throw new RouterException("Must provide at least one chunk for stitchBlob call",
+          RouterErrorCode.InvalidPutArgument);
+    }
```

Comment:
Yes, that is the idea.

I debated making the minimum number of chunks to stitch 2, but felt that this could be an unnecessary restriction in cases where someone thought they needed to perform a stitched upload, but only uploaded one chunk.

I also considered returning the original chunk blob ID in this case, but this makes the behavior inconsistent since the stitchBlob operation allows the user to specify new blob properties and user metadata for the composite blob and returning the original ID would retain the chunk's blob info.